codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ConstrainedCollectionGenerator.kt

```diff
@@ -0,0 +1,229 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package software.amazon.smithy.rust.codegen.server.smithy.generators
+
+import software.amazon.smithy.codegen.core.Symbol
+import software.amazon.smithy.model.shapes.CollectionShape
+import software.amazon.smithy.model.traits.LengthTrait
+import software.amazon.smithy.model.traits.Trait
+import software.amazon.smithy.rust.codegen.core.rustlang.Attribute
+import software.amazon.smithy.rust.codegen.core.rustlang.RustMetadata
+import software.amazon.smithy.rust.codegen.core.rustlang.RustWriter
+import software.amazon.smithy.rust.codegen.core.rustlang.Visibility
+import software.amazon.smithy.rust.codegen.core.rustlang.docs
+import software.amazon.smithy.rust.codegen.core.rustlang.documentShape
+import software.amazon.smithy.rust.codegen.core.rustlang.join
+import software.amazon.smithy.rust.codegen.core.rustlang.rust
+import software.amazon.smithy.rust.codegen.core.rustlang.rustTemplate
+import software.amazon.smithy.rust.codegen.core.smithy.RuntimeType
+import software.amazon.smithy.rust.codegen.core.util.PANIC
+import software.amazon.smithy.rust.codegen.core.util.orNull
+import software.amazon.smithy.rust.codegen.server.smithy.PubCrateConstraintViolationSymbolProvider
+import software.amazon.smithy.rust.codegen.server.smithy.ServerCodegenContext
+import software.amazon.smithy.rust.codegen.server.smithy.supportedCollectionConstraintTraits
+import software.amazon.smithy.rust.codegen.server.smithy.validationErrorMessage
+
+/**
+ * [ConstrainedCollectionGenerator] generates a wrapper tuple newtype holding a constrained `std::vec::Vec`.
+ * This type can be built from unconstrained values, yielding a `ConstraintViolation` when the input does not satisfy
+ * the constraints.
+ *
+ * The [`length`] and [`uniqueItems`] traits are the only constraint traits applicable to list shapes.
+ * The [`uniqueItems`] trait has not been implemented yet.
+ *
+ * If [unconstrainedSymbol] is provided, the `MaybeConstrained` trait is implemented for the constrained type, using the
+ * [unconstrainedSymbol]'s associated type as the associated type for the trait.
+ *
+ * [`length`]: https://smithy.io/2.0/spec/constraint-traits.html#length-trait
+ * [`uniqueItem`]: https://smithy.io/2.0/spec/constraint-traits.html#smithy-api-uniqueitems-trait
+ */
+class ConstrainedCollectionGenerator(
+    val codegenContext: ServerCodegenContext,
+    val writer: RustWriter,
+    val shape: CollectionShape,
+    private val constraintsInfo: List<TraitInfo>,
```

Comment:
nit: I think it'd be nicer if `ConstrainedCollectionGenerator` calculated `constraintsInfo` from the collection shape rather than relying on callers.

Reply:
I'd done it that way initially, but I moved it to a parameter because `CollectionConstraintViolationGenerator` needs it too, so I wanted to avoid computing it twice.